### ISSUE TITLE
[research] AERA-v26.10 post-#731 differential kernel-time attribution

### ISSUE THREAD

**Issue by vinceackermann2-sys** [vinceackermann2-sys/tam-research#732]
## Purpose

Explain the clean scientific/performance FAIL from #731 before attempting another optimization. The v26.10 latent-depth synchronization coalescing candidate preserved exact semantics and removed 19/21 `cudaStreamSynchronize` calls at batch 8/64, yet improved unprofiled full-call median latency by only 0.7608% / 0.6548%. Therefore synchronization-count reduction itself is not the dominant remaining wall-clock lever.

This issue is diagnostic-only. It authorizes no optimization implementation and no E2E systems rerun.

## Exact live source and parent result

Bind current source main/tree at preregistration:
- `SOURCE_MAIN=267f6d6018a0d355436869919ed73c7723315eab`
- `SOURCE_TREE=1739ccd80ad48636a4439cbc9d39ba1ab34df807`

Authoritative parent optimization attempt:
- research mechanism: #706 v26.10 latent-depth sync coalescing
- canonical L4 trigger: `#731`
- Actions evidence comment: `5574730812`
- workflow run `34154255434`
- job `101842659688`
- attempt `1`
- device `NVIDIA L4`
- frozen #710 memory-safe benchmark launcher blob `b885250753ea169cd89dd7a978bb3647fd261fe8`
- frozen v26.10 implementation blob `d8f691c198eed1fa96bcbb78a4e76cad82d18779`
- decision `FAIL_V26_10_LATENT_DEPTH_SYNC_COALESCING_MICROBENCH`
- `overall_pass=false`
- `fresh_full_e2e_systems_gate_authorized=false`

Exact #731 batch results:
- batch 8 baseline/candidate median `33.98348808288574 / 33.72492790222168 ms`; improvement `0.0076084061775363995`; sync `49 -> 30`, reduction `19`; correctness/route/decision/logits PASS, latency gate FAIL, sync gate FAIL.
- batch 64 baseline/candidate median `76.74127960205078 / 76.23878479003906 ms`; improvement `0.006547907653057826`; sync `51 -> 30`, reduction `21`; correctness/route/decision/logits PASS, latency gate FAIL, sync gate PASS.
- parameter schema/state-dict/checkpoint immutability PASS.

No #731 result may be rerun, retried, redispatched, or reinterpreted as a systems PASS.

## Historical exclusions

Do not repeat mechanisms already falsified as sufficient:

1. #405/#400 coalesced optional-state movement reduced CUDA device-event count to `0.533x` legacy for every measured row but was slower: batch64 latency geomean `1.115x`, with every row latency ratio >1.0. Launch-count reduction alone is therefore not evidence of latency benefit.
2. #258 reasoner-only `torch.compile` preserved exact selected-row semantics but delivered only about `1.09-1.12x`; compile-friendly eager was slightly faster than compiled. `torch.compile` alone is not the next systems solution.
3. #731 specifically falsified latent-depth synchronization coalescing as a >=5% full-call optimization.

## Diagnostic question

On the exact same frozen v26.9 baseline and exact frozen v26.10 candidate used by #731, what CUDA device work remains effectively unchanged after removing ~20 stream synchronizations, and which repeated kernel family or fusible operator cluster accounts for material wall-clock/device-active time at BOTH frozen batches?

The diagnostic must compare baseline and v26.10 side-by-side so the useful signal is the **differential timeline**, not raw launch count.

## Frozen execution fixtures

Reuse exactly:
- seed8471 checkpoints and hashes from #731, read-only;
- batches `(8,64)`;
- token seed rule inherited from #710/#706;
- `route_mode="hard_sparse"`, `hard=True`, `update_memory=True`;
- CUDA BF16 autocast through the frozen model call;
- exact v26.9 backend and exact v26.10 installer;
- no training, optimizer, backward, checkpoint write, corpus access, or new scientific seed.

Before profiling, require the same correctness gates used by #731: exact route/expert/depth decisions, exact active-row populations, parameter/state-dict schema identity, checkpoint hashes, finite outputs, logit/state equivalence under `atol=rtol=1e-2`, and no dense-masked sparse credit.

## Measurement boundary

Use low-intrusion PyTorch/Kineto CUDA timeline profiling. For each batch and each condition (`baseline_v26_9`, `candidate_v26_10`):
- 3 warmup calls;
- 3 separately measured unprofiled CUDA-event full-call calls to establish profiler distortion only;
- exactly 1 profiled whole call after timing;
- no profiler-derived timing may substitute for unprofiled latency evidence.

Persist at minimum:
- unprofiled median/mean/min/max full-call latency;
- profiled full-call CUDA-event latency and distortion ratio;
- chronological CUDA kernel/memcpy activity intervals within an explicit whole-call annotation;
- unioned device-active time;
- inter-device-activity idle-gap total/count/median/p95/max;
- total CUDA kernel count and memcpy count;
- kernel duration distribution: count and union/sum time for `<=10us`, `10-25us`, `25-50us`, `50-100us`, `>100us` buckets;
- top device kernel names by count and by total duration;
- for each top repeated name: count, total duration, median duration, p95 duration, and baseline-vs-v26.10 delta;
- exposed CUDA runtime launch/synchronize/memcpy counts;
- route/expert/depth/active-row evidence proving the same sparse geometry;
- explicit unavailable fields rather than inferred values when Kineto correlation is incomplete.

Do not infer kernel↔ATen ownership where trace correlation is absent. Raw CUDA kernel names/device durations may be reported directly.

## Preregistered interpretation / next-target rule

This diagnostic may name at most ONE `next_target`, and even then it creates no optimization authority automatically.

Candidate classes:
- `repeated_small_kernel_fusion::<kernel-or-stable-kernel-cluster>` only if the same named kernel/cluster is present at both batches, has at least `100` calls per full profiled call at both batches, and its distortion-normalized aggregate duration is at least `1.0 ms` AND at least `10%` of the parent unprofiled full-call median at both batches.
- `explicit_device_kernel_optimization::<kernel>` only if the same single kernel name is the largest device-time contributor at both batches and its distortion-normalized aggregate duration is at least `1.0 ms` AND `10%` of parent unprofiled latency at both batches.
- `whole_stage_launch_coalescing` only if no individual kernel/cluster meets the two rules above, but at both batches at least `500` kernel launches are `<=50us`, those small kernels account for at least `20%` of observed device-active duration, AND measured inter-activity idle gaps contribute at least `1.0 ms` after distortion normalization. Historical #405 means kernel-count reduction alone is insufficient for this classification.
- otherwise `next_target=None` and report batch-specific rankings.

For all distortion normalization, scale profiler component duration only by `unprofiled_median / profiled_cuda_event_latency`; never allow normalized components to exceed the unprofiled median, and label them diagnostic estimates rather than systems evidence.

A target must be common to batch 8 and 64. Batch-specific targets cannot authorize a shared optimization.

## Governance

CPU-first. This issue initially authorizes ZERO GPU work.

After issue number assignment:
1. fresh-audit live main/issues/PRs;
2. create only issue-specific diagnostic launcher/workflow/CPU-static-test blobs as detached Git objects;
3. freeze exact source main/tree, #731 lineage, #405/#258 historical exclusions, checkpoints, runtime/source blobs, result path, fresh trigger prefixes, thresholds, new blobs, detached tree, and detached commit in one comment before any branch/ref move;
4. only then create the named branch;
5. open one immutable PR and consume exactly one original full-repository CPU CI attempt; never rerun/retry/redispatch;
6. merge only the exact CPU-green frozen head after fresh live audit;
7. after merge: exactly one fresh zero-GPU preauth trigger, then one separate exact-main L4 authorization, then exactly one one-L4 diagnostic trigger.

Prefer exactly three additive files and explicitly package every root-level Modal dependency.

## Authority ceiling

This diagnostic cannot itself authorize an optimization, full E2E systems rerun, systems PASS, architecture freeze, S2, fresh scientific seed, replication credit, scaling, or breakthrough claim. Any non-null `next_target` requires a separately preregistered optimization successor. `next_target=None` means gather more evidence rather than guessing.

**Comment by vinceackermann2-sys** [comment]
## #732 pre-implementation differential kernel-time attribution freeze

Freeze before any #732 branch/ref move. Diagnostic-only successor after authoritative #731 clean microbenchmark FAIL. No optimization, E2E systems rerun, systems PASS, architecture freeze, S2, fresh scientific seed, replication, scaling, or breakthrough authority.

SOURCE_MAIN=267f6d6018a0d355436869919ed73c7723315eab
SOURCE_TREE=1739ccd80ad48636a4439cbc9d39ba1ab34df807
BRANCH=research/aera-v26-10-issue732-differential-kernel-time-attribution
PREAUTH_PREFIX=[aera-v26-10-issue732-differential-kernel-preauth]
L4_PREFIX=[aera-v26-10-issue732-differential-kernel-l4]
RESULT_PATH=/vol/aera-v26/issue732-post731-differential-kernel-time-attribution/result.json
AUTHORIZATION_HEADING=## #732 sole L4 differential kernel-time attribution authorization

PARENT_TRIGGER=731
PARENT_EVIDENCE_COMMENT=5574730812
PARENT_RUN=34154255434
PARENT_JOB=101842659688
PARENT_ATTEMPT=1
PARENT_DECISION=FAIL_V26_10_LATENT_DEPTH_SYNC_COALESCING_MICROBENCH
PARENT_BATCH8_BASELINE_MS=33.98348808288574
PARENT_BATCH8_CANDIDATE_MS=33.72492790222168
PARENT_BATCH8_SYNC=49,30
PARENT_BATCH64_BASELINE_MS=76.74127960205078
PARENT_BATCH64_CANDIDATE_MS=76.23878479003906
PARENT_BATCH64_SYNC=51,30

HISTORICAL_EXCLUSION_405=launch-count/coalesced optional-state movement reduced events but was slower; do not infer latency benefit from launch-count reduction alone.
HISTORICAL_EXCLUSION_258=reasoner-only torch.compile was insufficient; do not repeat compile-alone as the systems solution.

ISSUE710_LAUNCHER_BLOB=b885250753ea169cd89dd7a978bb3647fd261fe8
V26_10_IMPL_BLOB=d8f691c198eed1fa96bcbb78a4e76cad82d18779
CHECKPOINT_AERA_SHA256=f8aa92421801e8f190247e420632be5f0c20bc5ea8bf6bdeefe06686b3a31b30
CHECKPOINT_TRANSFORMER_SHA256=cdd5cab4439a709468d6607d45d82081b33e876b2e40d91d4a38ba139b219dd7

BATCHES=8,64
WARMUP_CALLS=3
UNPROFILED_CALLS=3
PROFILE_CALLS=1
REPEATED_MIN_CALLS=100
REPEATED_MIN_MS=1.0
REPEATED_MIN_SHARE=0.10
WHOLE_STAGE_MIN_SMALL_CALLS=500
WHOLE_STAGE_MIN_SMALL_ACTIVE_SHARE=0.20
WHOLE_STAGE_MIN_IDLE_MS=1.0
INTEGRATED_ATOL=0.01
INTEGRATED_RTOL=0.01

LAUNCHER_BLOB=c0ed30de101c890f812d5b0fef0047056d81f9fb
WORKFLOW_BLOB=193e03b429ae3829fb67fc2e1fe786c7ca9529f1
CPU_TEST_BLOB=a4f57b0bf79cdfc4ee2b8f122939539b5aec813d
FROZEN_TREE=fe5181034b3cb07076b17a43c7137829124d8cbb
FROZEN_COMMIT=311190f0dc2b0d584fb2ce5f83b064aabc6493ab

Exact detached diff from SOURCE_MAIN: exactly three additive files, 556 additions, 0 deletions: one Modal diagnostic launcher, one owner-only issues-opened workflow, one CPU/static contract test. No existing runtime/model/checkpoint/evaluator file is modified.

Preregistered decision classes remain exactly: `repeated_small_kernel_fusion::<kernel>`, `explicit_device_kernel_optimization::<kernel>`, `whole_stage_launch_coalescing`, or `None`, subject to the thresholds above and common evidence at both frozen batches. Profiler component durations are distortion-normalized by unprofiled_median/profiled_cuda_event_latency and capped at the unprofiled median; profiler timing is diagnostic only.

Only after this freeze may the named branch be created at exactly FROZEN_COMMIT, followed by one immutable PR and exactly one original full-repository CPU CI attempt. Never rerun/retry/redispatch a consumed attempt.

**Comment by vinceackermann2-sys** [comment]
## #732 sole L4 differential kernel-time attribution authorization

Authorize main: `d15d94f39e59d0c6cc8d1514f44a33452d279a74`

Prerequisites are satisfied exactly:
- frozen #732 diagnostic tree `fe5181034b3cb07076b17a43c7137829124d8cbb`, freeze comment `5574897860`;
- merged PR #733 exact frozen head `311190f0dc2b0d584fb2ce5f83b064aabc6493ab`;
- sole original PR CPU CI run `34155832359`, job `101847350756`, attempt `1`: `1149 passed, 15 warnings`;
- successful sole zero-GPU preauth trigger #734;
- preauth evidence comment `5574961951`, proving result absent and revalidating #731 durable result, checkpoints, frozen #710 launcher, and v26.10 implementation with no GPU/model/new measurement.

Authorize exactly one fresh issue under frozen L4 prefix `[aera-v26-10-issue732-differential-kernel-l4]` on the exact main above. Execute only the frozen #732 diagnostic launcher on one NVIDIA L4 with the preregistered batches, seeds, correctness gates, three unprofiled calls per condition and one Kineto call per condition.

Purpose only: differential kernel/device-time attribution after #731 FAIL. This authorization grants no optimization implementation, E2E systems rerun, systems PASS, architecture freeze, S2, fresh scientific seed, replication, scaling, or breakthrough claim. The future L4 trigger/attempt is consumed regardless of outcome and must never be rerun/retried/redispatched.
